request-manager/graph/resolver.go

```diff
@@ -121,193 +123,168 @@ func (o *resolver) RequestArgs(jobArgs map[string]interface{}) ([]proto.RequestA
+// a job, buildSequence creates the corresponding job. If a sequence graph node needs
+// to be expanded, i.e. it represents anything but a job, it is recursively expanded
+// with another call to buildSequence. buildSequence also chooses the correct path
+// for conditional nodes.
+func (r *resolver) buildSequence(cfg buildSequenceConfig) (*Graph, error) {
 	sequenceName := cfg.sequenceName
 	jobArgs := cfg.jobArgs
 
-	/* Add optional and static sequence arguments to map. */
-	seq, ok := o.AllSequences[sequenceName]
-	if !ok {
-		return nil, fmt.Errorf("cannot find definition for sequence: %s", sequenceName)
-	}
-	for _, arg := range seq.Args.Optional {
-		if _, ok := jobArgs[*arg.Name]; !ok {
-			jobArgs[*arg.Name] = *arg.Default
-		}
-	}
-	for _, arg := range seq.Args.Static {
-		if _, ok := jobArgs[*arg.Name]; !ok {
-			jobArgs[*arg.Name] = *arg.Default
-		}
-	}
-
-	/* Build request graph based on sequence graph. */
-	seqGraph, ok := o.SequenceGraphs[sequenceName]
+	// Build request graph based on sequence graph. We use the sequence graph
+	// as a template, traversing it in topological order and processing each
+	// of its nodes depending on what category it is (job, sequence, conditional).
+	// This sequence graph won't be mutated at all.
+	seqGraph, ok := r.sequenceGraphs[sequenceName]
 	if !ok {
+		// Graph checks should prevent this from happening.
+		// If this error is thrown, there's a bug in the code.
 		return nil, fmt.Errorf("cannot find sequence graph for sequence: %s", sequenceName)
 	}
 
-	reqGraph, err := o.newReqGraph(cfg.wrapperName, jobArgs)
+	// This is the graph we'll actually be filling out. It starts off empty,
+	// and we'll add nodes as we traverse the sequence graph.
+	reqGraph, err := r.newReqGraph(cfg.graphName, jobArgs)
 	if err != nil {
 		return nil, err
 	}
 
-	// Traverse sequence graph in topological order. Each sequence graph node corresponds to either a single
-	// job node, or an entire subgraph of nodes. Build the subgraph and insert it into the larger sequence
-	// graph.
+	// Traverse sequence graph in topological order. Each sequence graph node
+	// corresponds to either a single job node, or an entire subgraph of nodes.
+	// Build the subgraph and insert it into the larger sequence graph.
 	idMap := map[string]*Graph{} // sequence graph node id --> corresponding subgraph
 	for _, seqGraphNode := range seqGraph.Order {
-		n := seqGraphNode.Spec
-
-		// Find out how many times this node has to be repeated
-		elements, lists, err := o.getIterators(n, jobArgs)
+		nodeSpec := seqGraphNode.Spec
+
+		// 1. Do sequence expansion.
+		// Find out how many times this node has to be repeated.
+		// If this node has a `each:` field, then lists[i] and
+		// elements[i] describe the ith `list:element`. Specifically,
+		// lists[i] is the actual list value, not just the name of the
+		// job arg.
+		elements, lists, err := r.getIterators(nodeSpec, jobArgs)
```

Comment:
👍  Nice. I'm still not a big fan of list:element, but the code comment helps so much that I'm ok with it now. I have something to refer back to when I forget what list and element actually are/contain.